### PR TITLE
Add `backends!` and `backends_map!` macros to code duplication

### PR DESCRIPTION
**Pull request by Kimundi**
**Description**
The source code is doing a lot of things duplicated for each backend, which leads to code duplication and harder to read code.
This PR attempts to improve the situation by introducing a `backends!` and `backends_map!` for doing the code duplication automatically.

**Testing**
The macro have unit tests for checking that they work correctly.

### CODE REVIEW

File: wgpu-core/src/instance.rs [hunk 1]
```diff
@@ -59,61 +59,45 @@ pub struct Instance {
...
+            let map = |(backend, backend_create)| {
+                if backends.contains(backend.into()) {
+                    backend_create()
+                } else {
+                    None
+                }
+            };
+            Instance {
+                #[vulkan]
+                vulkan: map((Backend::Vulkan, || gfx_backend_vulkan::Instance::create(name, version).ok())),
```
Comment (kvark) [inline wgpu-core/src/instance.rs:72]: can we just pass `gfx_backend_vulkan::Instance::create` here? It's a static function with known parameters, and map() implementation can always just call it as `backend_crate(name, version).ok()`

File: wgpu-core/src/instance.rs [hunk 2]
```diff
@@ -555,64 +453,39 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
+                        adapter,
+                        &mut token,
+                    );
+                    return Some(id);
+                }
+                selected -= adapters_backend.len();
+            };
+
+            #[vulkan]
+            map(("Vulkan", &mut id_vulkan, adapters_vk, || backend::Vulkan::hub(self))),
```
Comment (kvark) [inline wgpu-core/src/instance.rs:473]: can we pass it as `backend::Vulkan::hub`? again, it's a static function with known argument, which `map()` can call on `self`

File: wgpu-core/src/macros.rs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+            #[cfg(windows)]
+            $dx12_field: $dx12_expr,
+
+            #[cfg(windows)]
+            $dx11_field: $dx11_expr,
+        }
+    }};
+}
+
+#[allow(unused_macros)]
```
Comment (kvark) [inline wgpu-core/src/macros.rs:80]: why would it be unused? it's used internally

File: wgpu-core/src/instance.rs (hunk 1 above)
Comment (Kimundi) [reply wgpu-core/src/instance.rs:72]: So far I've preserved the semantic of this function where some of the backends panic with unwrap(), and some just silently map to None. By passing `create`, we would have to uniformly do one or the other here. If that is fine I can change it. 😄

File: wgpu-core/src/instance.rs (hunk 2 above)
Comment (Kimundi) [reply wgpu-core/src/instance.rs:473]: Sure, I'll change it.

File: wgpu-core/src/macros.rs (hunk 3 above)
Comment (Kimundi) [reply wgpu-core/src/macros.rs:80]: Whoops, artifact from development

File: wgpu-core/src/device/mod.rs [hunk 4]
```diff
@@ -2671,17 +2671,20 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
-            feature = "gfx-backend-vulkan"
-        ))]
-        self.poll_devices::<backend::Vulkan>(force_wait, &mut callbacks);
-        #[cfg(windows)]
-        self.poll_devices::<backend::Dx11>(force_wait, &mut callbacks);
-        #[cfg(windows)]
-        self.poll_devices::<backend::Dx12>(force_wait, &mut callbacks);
-        #[cfg(any(target_os = "ios", target_os = "macos"))]
-        self.poll_devices::<backend::Metal>(force_wait, &mut callbacks);
+        backends! {
```
Comment (Kimundi) [inline wgpu-core/src/device/mod.rs:2674]: One thing I forgot when I posted this PR is that this is the only place where the order of statements changed with the macros. I assume this does not make a difference, but can you verify? :)

Comment (kvark) [reply wgpu-core/src/device/mod.rs:2674]: correct, it's fine to reorder

File: wgpu-core/src/instance.rs (hunk 1 above)
Comment (kvark) [reply wgpu-core/src/instance.rs:72]: yeah, that's fine, let's have all of them this way

**Pull request merged**